File: messconnect-api/src/main/java/com/messconnect/api/domain/Subscription.java
package com.messconnect.api.domain;

import com.messconnect.api.domain.enums.BillingCycle;
import com.messconnect.api.domain.enums.MealType;
import com.messconnect.api.domain.enums.SubscriptionStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * A user's subscription to a mess plan (PART 9).
 */
@Entity
@Table(name = "subscriptions")
public class Subscription {

	@Id
	@GeneratedValue(strategy = GenerationType.UUID)
	private UUID id;

	@Column(name = "user_id", nullable = false)
	private UUID userId;

	@Column(name = "mess_id", nullable = false)
	private UUID messId;

	@Column(name = "plan_name", nullable = false)
	private String planName;

	@Enumerated(EnumType.STRING)
	@Column(name = "meal_type", nullable = false)
	private MealType mealType;

	@Enumerated(EnumType.STRING)
	@Column(name = "billing_cycle", nullable = false)
	private BillingCycle billingCycle;

	@Column(name = "price", nullable = false)
	private BigDecimal price = BigDecimal.ZERO;

	@Enumerated(EnumType.STRING)
	@Column(name = "status", nullable = false)
	private SubscriptionStatus status = SubscriptionStatus.ACTIVE;

	@Column(name = "start_date", nullable = false)
	private LocalDate startDate = LocalDate.now();

	@Column(name = "end_date")
	private LocalDate endDate;

	@Column(name = "created_at", nullable = false, updatable = false)
	private Instant createdAt = Instant.now();

	public UUID getId() {
		return id;
	}

	public void setId(UUID id) {
		this.id = id;
	}

	public UUID getUserId() {
		return userId;
	}

	public void setUserId(UUID userId) {
		this.userId = userId;
	}

	public UUID getMessId() {
		return messId;
	}

	public void setMessId(UUID messId) {
		this.messId = messId;
	}

	public String getPlanName() {
		return planName;
	}

	public void setPlanName(String planName) {
		this.planName = planName;
	}

	public MealType getMealType() {
		return mealType;
	}

	public void setMealType(MealType mealType) {
		this.mealType = mealType;
	}

	public BillingCycle getBillingCycle() {
		return billingCycle;
	}

	public void setBillingCycle(BillingCycle billingCycle) {
		this.billingCycle = billingCycle;
	}

	public BigDecimal getPrice() {
		return price;
	}

	public void setPrice(BigDecimal price) {
		this.price = price;
	}

	public SubscriptionStatus getStatus() {
		return status;
	}

	public void setStatus(SubscriptionStatus status) {
		this.status = status;
	}

	public LocalDate getStartDate() {
		return startDate;
	}

	public void setStartDate(LocalDate startDate) {
		this.startDate = startDate;
	}

	public LocalDate getEndDate() {
		return endDate;
	}

	public void setEndDate(LocalDate endDate) {
		this.endDate = endDate;
	}

	public Instant getCreatedAt() {
		return createdAt;
	}

	public void setCreatedAt(Instant createdAt) {
		this.createdAt = createdAt;
	}
}

File: messconnect/lib/features/subscriptions/models/subscription.dart
/// Client-side model mirroring the backend [SubscriptionResponse] (PART 9).
class Subscription {
  const Subscription({
    required this.id,
    required this.userId,
    required this.messId,
    required this.planName,
    required this.mealType,
    required this.billingCycle,
    required this.price,
    required this.status,
    required this.startDate,
    required this.endDate,
    required this.createdAt,
  });

  final String id;
  final String userId;
  final String messId;
  final String planName;
  final String mealType;
  final String billingCycle;
  final double price;
  final String status;
  final DateTime startDate;
  final DateTime endDate;
  final DateTime createdAt;

  factory Subscription.fromJson(Map<String, dynamic> json) {
    return Subscription(
      id: json['id'] as String,
      userId: json['userId'] as String,
      messId: json['messId'] as String,
      planName: json['planName'] as String,
      mealType: json['mealType'] as String,
      billingCycle: json['billingCycle'] as String,
      price: (json['price'] as num).toDouble(),
      status: json['status'] as String,
      startDate: DateTime.parse(json['startDate'] as String),
      endDate: DateTime.parse(json['endDate'] as String),
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  bool get isActiveOrPrevious =>
      status == 'ACTIVE' || status == 'EXPIRED' || status == 'CANCELLED';
}